### PR TITLE
add support for f16

### LINKED ISSUE #890: add support for half precision floats (f16)
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

a challenge: to include support of half precision floats (f16)

**Describe the solution you'd like**

the support can be optional, with the inclusion of `half` crate, which just wraps u16 with correct arithmatics

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

Closes #890 

# Rationale for this change
 
float16 was not properly supported

# What changes are included in this PR?

include optional feature `f16` to use `half` crate to implement f16

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/datatypes/numeric.rs [hunk 1]
```diff
@@ -333,6 +335,8 @@ make_numeric_type!(UInt8Type, u8, u8x64, m8x64);
 make_numeric_type!(UInt16Type, u16, u16x32, m16x32);
 make_numeric_type!(UInt32Type, u32, u32x16, m32x16);
 make_numeric_type!(UInt64Type, u64, u64x8, m64x8);
+#[cfg(feature = "f16")]
+make_numeric_type!(Float16Type, f16, f16x32, m16x32);
```
Comment (Jimexist) [inline arrow/src/datatypes/numeric.rs:339]: I wish there were support for this but apparently there isn't

File: arrow/src/json/reader.rs [hunk 2]
```diff
@@ -1015,7 +1015,15 @@ impl Decoder {
             DataType::UInt32 => self.read_primitive_list_values::<UInt32Type>(rows),
             DataType::UInt64 => self.read_primitive_list_values::<UInt64Type>(rows),
             DataType::Float16 => {
-                return Err(ArrowError::JsonError("Float16 not supported".to_string()))
+                #[cfg(feature = "f16")]
```
Comment (alamb) [inline arrow/src/json/reader.rs:1018]: this seems redundant -- both branches do the same thing. Was this an oversight? Or perhaps just a future TODO that will be clearer when separated like this?

Comment (alamb) [reply arrow/src/json/reader.rs:1018]: Same question applies below

**Review by alamb — COMMENTED** [review]
Looks pretty cool to me. I am not sure how widely used the `Float16` type is but seems like a reasonable feature to add

I think it would be good to add some basic tests for `Float16Array` (like, for example, creating it from an iterator) as well as an example as a doc comment).

I also was wondering if there was some way to avoid the sprinkled `#[cfg(feature = "f16")]` throughout the code (for example, can we possible always define `Float16Array` even if the `f16` type is not defined, but provide some implementation that always errors when it is constructed.

Then we could have one or two `#[cfg(feature = "f16")]` checks rather than them throughout the code

File: arrow/src/array/data.rs [hunk 3]
```diff
@@ -88,6 +89,19 @@ pub(crate) fn new_buffers(data_type: &DataType, capacity: usize) -> [MutableBuff
...
+            #[cfg(feature = "f16")]
+            {
+                [
+                    MutableBuffer::new(capacity * mem::size_of::<f16>()),
+                    empty_buffer,
+                ]
+            }
+            #[cfg(not(feature = "f16"))]
+            {
+                unimplemented!()
```
Comment (martin-g) [inline arrow/src/array/data.rs:102]: `unimplemented!("Float16 datatype not supported")` as in the other places

**Comment by Jimexist** [comment]
> 

> can we possible always define Float16Array even if the f16 type is not defined, but provide some implementation that always errors when it is constructed.

are you saying that we should allow for runtime error rather than compile time failures?

**Comment by Jimexist** [comment]
> Looks pretty cool to me. I am not sure how widely used the `Float16` type is but seems like a reasonable feature to add
> 
> I think it would be good to add some basic tests for `Float16Array` (like, for example, creating it from an iterator) as well as an example as a doc comment).
> 
> I also was wondering if there was some way to avoid the sprinkled `#[cfg(feature = "f16")]` throughout the code (for example, can we possible always define `Float16Array` even if the `f16` type is not defined, but provide some implementation that always errors when it is constructed.
> 
> Then we could have one or two `#[cfg(feature = "f16")]` checks rather than them throughout the code

> can we possible always define Float16Array even if the f16 type is not defined, but provide some implementation that always errors when it is constructed.

are you saying that we should allow for runtime error rather than compile time failures?

**Review by alamb — COMMENTED** [review]
Looks pretty cool to me -- thanks @Jimexist 

One major questions for other reviewers:  Are we ok with adding a new dependency (on `half`?) -- in the IOx project, it turns out we already have `half` because `serde_cbor` depends on it, because `criterion depends on it` 🤷 

I think some basic tests are in order -- perhaps a doctest showing how to construct an `F16Array` perhaps?

**Comment by Jimexist** [comment]
> Looks pretty cool to me -- thanks @Jimexist
> 
> One major questions for other reviewers: Are we ok with adding a new dependency (on `half`?) -- in the IOx project, it turns out we already have `half` because `serde_cbor` depends on it, because `criterion depends on it` 🤷
> 
> I think some basic tests are in order -- perhaps a doctest showing how to construct an `F16Array` perhaps?

do you mean this: https://github.com/apache/arrow-rs/pull/888/files#diff-dd779d74625591eb0e2a5648a76db5b714f4135ee094afd77e87a584f9f264fbR199?

File: arrow/src/array/mod.rs [hunk 4]
```diff
@@ -192,6 +192,14 @@ pub type UInt64Array = PrimitiveArray<UInt64Type>;
 ///
 /// # Example: Using `collect`
 /// ```
+/// # use arrow::array::Float16Array;
+/// use half::f16;
+/// let arr : Float16Array = [Some(f16::from_f64(1.0)), Some(f16::from_f64(2.0))].into_iter().collect();
```
Comment (alamb) [inline arrow/src/array/mod.rs:197]: ❤️

**Review by alamb — APPROVED** [review]
Looks good -- thanks @Jimexist

**Comment by Jimexist** [comment]
can I possibly get another stamp on this? I guess introducing `half` isn't going to be obviously all upside

**Review by alamb — APPROVED** [review]
Sorry -- I meant to approve this earlier.

**Comment by Jimexist** [comment]
@alamb with no further reviews and concerns i have merged this pull request

**Comment by alamb** [comment]
Nice work @Jimexist  👍

**Pull request merged**